security/class-user-last-seen.php

```diff
@@ -0,0 +1,420 @@
+				'meta_value'   => $this->get_inactivity_timestamp(),
+				'meta_type'    => 'NUMERIC',
+				'meta_compare' => '<',
+				'count_total'  => false,
+				'number'       => 1, // To minimize the query time, we only need to know if there are any blocked users to show the link
+			),
+		);
+
+		$views['blocked_users'] = __( 'Blocked Users', 'wpvip' );
+
+		if ( ! $users_query->get_results() ) {
+			return $views;
+		}
+
+		$url = add_query_arg( array(
+			'last_seen_filter'       => 'blocked',
+			'last_seen_filter_nonce' => wp_create_nonce( 'last_seen_filter' ),
+		) );
+
+		// phpcs:ignore WordPress.Security.NonceVerification.Recommended
+		$class = isset( $_GET['last_seen_filter'] ) ? 'current' : '';
+
+		$view = '<a class="' . esc_attr( $class ) . '" href="' . esc_url( $url ) . '">' . esc_html( $views['blocked_users'] ) . '</a>';
+
+		$views['blocked_users'] = $view;
+
+		return $views;
+	}
+
+	public function last_seen_unblock_action() {
+		$admin_notices_hook_name = is_network_admin() ? 'network_admin_notices' : 'admin_notices';
+
+		if ( isset( $_GET['reset_last_seen_success'] ) && '1' === $_GET['reset_last_seen_success'] ) {
+			add_action( $admin_notices_hook_name, function () {
+				$class = 'notice notice-success is-dismissible';
+				$error = __( 'User unblocked.', 'wpvip' );
+
+				printf( '<div class="%1$s"><p>%2$s</p></div>', esc_attr( $class ), esc_html( $error ) );
+			} );
+		}
+
+		if ( ! isset( $_GET['user_id'], $_GET['action'] ) || 'reset_last_seen' !== $_GET['action'] ) {
+			return;
+		}
+
+		$user_id = absint( $_GET['user_id'] );
+
+		$error = null;
+		if ( ! isset( $_GET['reset_last_seen_nonce'] ) || ! wp_verify_nonce( sanitize_text_field( $_GET['reset_last_seen_nonce'] ), 'reset_last_seen_action' ) ) {
+			$error = __( 'Unable to verify your request', 'wpvip' );
+		}
+
+		if ( ! get_userdata( $user_id ) ) {
+			$error = __( 'User not found.', 'wpvip' );
+		}
+
+		if ( ! current_user_can( 'edit_user', $user_id ) ) {
+			$error = __( 'You do not have permission to unblock this user.', 'wpvip' );
+		}
+
+		$ignore_inactivity_check_until = strtotime( '+2 days' );
+		if ( ! $error && ! $this->ignore_inactivity_check_for_user( $user_id, $ignore_inactivity_check_until ) ) {
+			$error = __( 'Unable to unblock user.', 'wpvip' );
+		}
+
+		if ( $error ) {
+			add_action( $admin_notices_hook_name, function () use ( $error ) {
+				$class = 'notice notice-error is-dismissible';
+
+				printf( '<div class="%1$s"><p>%2$s</p></div>', esc_attr( $class ), esc_html( $error ) );
+			} );
+			return;
+		}
+
+		$url = remove_query_arg( array(
+			'action',
+			'user_id',
+			'reset_last_seen_nonce',
+		) );
+
+		$url = add_query_arg( array(
+			'reset_last_seen_success' => 1,
+		), $url );
+
+		wp_safe_redirect( $url );
+		exit();
+	}
+
+	public function ignore_inactivity_check_for_user( $user_id, $until_timestamp = null ) {
+		if ( ! $until_timestamp ) {
+			$until_timestamp = strtotime( '+2 days' );
```

Comment:
The default of `+2 days` feels a little too magic and is hard to predict. Maybe we should just require `$until_timestamp` be specified?

Reply:
Minor, but I'd also consider naming the function `ignore_inactivity_checek_for_user_until( $user_id, $until_timestamp )`. The `_until` makes it obvious that it's a limited time action, vs. permanent (which is what I thought it would do before I read the code)